src/tools/dtrust-tool.c

```diff
@@ -197,29 +333,470 @@ unlock_transport_protection(sc_card_t *card)
 		printf("Can't change pin: %s\n", sc_strerror(r));
 
 fail:
-	if (qespin2 != NULL) {
-		sc_mem_clear(qespin2, qespin2_len);
-		free(qespin2);
+	if (qespin != NULL) {
+		sc_mem_clear(qespin, strlen(qespin));
+		free(qespin);
+	}
+
+	if (tpin != NULL) {
+		sc_mem_clear(tpin, strlen(tpin));
+		free(tpin);
+	}
+}
+
+void
+unlock_transport_protection5(sc_card_t *card, int ref_pace, int ref_pin, const char *pathstr, const char *pin_label)
+{
+	int r;
+	sc_path_t path;
+	struct sc_pin_cmd_data data;
+	char *tpin = NULL;
+	char *newpin = NULL;
+	int tries_left;
+
+	printf("Unlocking %s\n", pin_label);
+
+	/* Query all PINs at once */
+	if (!(card->reader->capabilities & SC_READER_CAP_PACE_GENERIC)) {
+		r = get_pin(&tpin, "Transport PIN", 0);
+		if (r < 0)
+			goto fail;
+	}
+
+	if (!(card->reader->capabilities & SC_READER_CAP_PIN_PAD)) {
+		r = get_pin(&newpin, pin_label, 1);
+		if (r < 0)
+			goto fail;
+
+		if (strlen(newpin) != 8) {
+			fprintf(stderr, "Error. New PIN must be exactly 8 characters long.\n");
+			goto fail;
+		}
+	}
+
+	/* Authenticate via PACE */
+	memset(&data, 0, sizeof(data));
+	data.cmd = SC_PIN_CMD_VERIFY;
+	data.pin_type = SC_AC_CHV;
+	data.pin_reference = ref_pace;
+
+	if (card->reader->capabilities & SC_READER_CAP_PACE_GENERIC)
+		printf("Enter Transport PIN on the readers pin pad now.\n");
+	else {
+		data.pin1.data = (u8 *)tpin;
+		data.pin1.len = strlen(tpin);
+	}
+
+	r = sc_pin_cmd(card, &data, &tries_left);
+	if (r) {
+		fprintf(stderr, "Error verifying Transport PIN: %s\n", sc_strerror(r));
+		if (tries_left >= 0)
+			fprintf(stderr, "%d attempts left.\n", tries_left);
+		goto fail;
+	}
+
+	/* Select application of the PIN */
+	sc_format_path(pathstr, &path);
+	r = sc_select_file(card, &path, NULL);
+	if (r)
+		goto fail;
+
+	/* Change PIN */
+	memset(&data, 0, sizeof(data));
+	data.cmd = SC_PIN_CMD_CHANGE;
+	data.flags = SC_PIN_CMD_IMPLICIT_CHANGE;
+	data.pin_type = SC_AC_CHV;
+	data.pin_reference = ref_pin;
+	data.pin2.min_length = 8;
+	data.pin2.max_length = 8;
+
+	if (card->reader->capabilities & SC_READER_CAP_PIN_PAD) {
+		printf("Enter new %s on the readers pin pad now.\n", pin_label);
+		data.pin2.prompt = pin_label;
+		data.flags |= SC_PIN_CMD_USE_PINPAD;
+	} else {
+		data.pin2.data = (u8 *)newpin;
+		data.pin2.len = strlen(newpin);
 	}
 
-	if (qespin1 != NULL) {
-		sc_mem_clear(qespin1, qespin1_len);
-		free(qespin1);
+	/* We only have one chance to set the new PIN. Once the Transport PIN
+	 * is verified, it is not usable anymore. For pin pad readers we
+	 * continue as long as the new PIN is set successfully or the user
+	 * aborts the program and renders its card unusable as a consequence. */
+	do {
+		r = sc_pin_cmd(card, &data, NULL);
+		if (r == SC_SUCCESS) {
+			printf("Transport protection removed. You can now use your %s.\n", pin_label);
+			break;
+		}
```

Comment:
This is an endless loop, if the user always aborts the insertion of the PIN

Reply:
This fail-operational pattern is intentional. When verifying the Transport PIN its use counter is set and cannot be used anymore. But the signature PIN still need to be changed. So whatever happens, the user should be prevented from accidentally aborting the program at this stage as it virtually bricks the card.

Nevertheless the user always has the possibility to abort the program by CTRL+C.